### DISCUSSION TITLE
CGAL to adopt smart pointers from STL

### DISCUSSION THREAD

**Discussion by szobov** [CGAL/cgal#5713] [Ideas]
Hi there,
First of all, thanks a lot for the awesome project!

I have a question, are there any plans to adopt smart pointers from the STL?

As far as I can understand, CGAL mostly tries to manage memory by direct calls of `new`/`delete`, which is quite error-prone. I've also noticed some places, where CGAL uses homegrown "smart" ref-counter, but even they can introduce new bugs, like use-after-free or double-free.

Shared and unique pointers are the best options to replace all this staff and reduce the number of `segfaults` issues in CGAL.
So, what the CGAL's team think about it?

Thanks for any info,
Cheers

**Comment by lrineau** [comment]
Where to begin with? In which of the CGAL packages do you think a use of small pointers might be beneficial?

In CGAL we certainly do not see usage of `new`/`delete` as a pattern to follow. The code you are talking about must be old.

**Reply by szobov** [reply]
Hi @lrineau 
Try to search for `new` or `delete` in these files:
https://github.com/CGAL/cgal/blob/master/STL_Extension/include/CGAL/Handle_for.h
https://github.com/CGAL/cgal/blob/master/Nef_3/include/CGAL/Nef_3/SNC_point_locator.h
https://github.com/CGAL/cgal/blob/master/STL_Extension/include/CGAL/Object.h
https://github.com/CGAL/cgal/blob/master/Nef_3/include/CGAL/Nef_polyhedron_3.h
(just to name a few)

**Reply by lrineau** [reply]
- `Object` is a legacy class from CGAL, that we are trying to remove from CGAL.
- `Handle_for<T>` is a smart pointer itself, and we already thinking about using a `std::shared_ptr` instead.
- Nef_3 is indeed one the CGAL packages that would need a refresh.

**Reply by szobov** [reply]
Happy to hear you already have it in mind!
Btw, if you'll have a plan/issues with a bit detailed description somewhere on Github, like on the "Projects" tab, it could be good stuff for the future contributors to help you with developing.
Anyway, thanks for the explanation.
Cheers
